### PR TITLE
[Merged by Bors] - feat(NumberTheory/LSeries): Even Hurwitz zeta functions (II)

### PR DESCRIPTION
**Pull request by loefflerd**
This is a sequel to #12147, adding the key properties of the even Hurwitz zeta function: it is differentiable away from s = 1, and it agrees with the Dirichlet series when the latter converges. 

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by loefflerd** [comment]
For some reason the build is failing on CI, although it works fine on my system. Will investigate tomorrow.

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 1]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+`s = 0` if `a ≠ 0`)
+-/
+lemma differentiableAt_completedHurwitzZetaEven
+    (a : UnitAddCircle) {s : ℂ} (hs : s ≠ 0 ∨ a ≠ 0) (hs' : s ≠ 1) :
+    DifferentiableAt ℂ (completedHurwitzZetaEven a) s := by
+  refine (((hurwitzEvenFEPair a).differentiableAt_Λ ?_ (Or.inl ?_)).comp s
+      (differentiableAt_id.div_const _)).div_const _
+  · rw [div_ne_zero_iff, eq_true_intro two_ne_zero, and_true]
+    refine Or.imp (by tauto) (fun ha ↦ ?_) hs
+    simp only [hurwitzEvenFEPair, eq_false_intro ha, if_false]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:428]:
```suggestion
  · simp only [ne_eq, div_eq_zero_iff, OfNat.ofNat_ne_zero, or_false]
    rcases hs with h | h
    · exact Or.inl h
    · simp only [hurwitzEvenFEPair, one_div, h, ↓reduceIte, or_true]
```
I think this reads a bit better (I find `eq_true_intro` and `eq_false_intro` somewhat suspicous).

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 2]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+  rw [funext this]
+  refine .sub ?_ <| (differentiable_const _ _).div (differentiable_id _) one_ne_zero
+  apply DifferentiableAt.sub <;> apply differentiable_completedHurwitzZetaEven₀
+
+lemma differentiableAt_completedCosZeta
+    (a : UnitAddCircle) {s : ℂ} (hs : s ≠ 0) (hs' : s ≠ 1 ∨ a ≠ 0) :
+    DifferentiableAt ℂ (completedCosZeta a) s := by
+  refine (((hurwitzEvenFEPair a).symm.differentiableAt_Λ (Or.inl ?_) ?_).comp s
+      (differentiableAt_id.div_const _)).div_const _
+  · rwa [div_ne_zero_iff, eq_true_intro two_ne_zero, and_true]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:455]:
```suggestion
  · exact div_ne_zero_iff.mpr ⟨hs, two_ne_zero⟩
```

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 3]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+    Differentiable ℂ (completedCosZeta₀ a) :=
+  ((hurwitzEvenFEPair a).symm.differentiable_Λ₀.comp (differentiable_id.div_const _)).div_const _
+
+/-- The residue of `completedHurwitzZetaEven a s` at `s = 1` is equal to `1`. -/
+lemma completedHurwitzZetaEven_residue_one (a : UnitAddCircle) :
+    Tendsto (fun s ↦ (s - 1) * completedHurwitzZetaEven a s) (𝓝[≠] 1) (𝓝 1) := by
+  have h1 : Tendsto (fun s : ℂ ↦ (s - ↑(1  / 2 : ℝ)) * _) (𝓝[≠] ↑(1  / 2 : ℝ))
+    (𝓝 ((1 : ℂ) * (1 : ℂ))) := (hurwitzEvenFEPair a).Λ_residue_k
+  simp only [push_cast, one_mul] at h1
+  have h2 : Tendsto (fun s : ℂ ↦ s / 2) (𝓝[≠] 1) (𝓝[≠] (1 / 2)) :=
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:471]: A version of this occurs three times here. Maybe add a
```lean
lemma xyz {a : ℂ} (b : ℂ) (ha : a ≠ 0) : Tendsto (fun s : ℂ ↦ s / a) (𝓝[≠] b) (𝓝[≠] (b / a))
```
somewhere? (Or state it as a private lemma for `a = 2`.)

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 4]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+/-- Formula for `completedCosZeta` as a Dirichlet series in the convergence range
+(first version, with sum over `ℤ`). -/
+lemma hasSum_int_completedCosZeta (a : ℝ) {s : ℂ} (hs : 1 < re s) :
+    HasSum (fun n : ℤ ↦ Gammaℝ s * cexp (2 * π * I * a * n) / (↑|n| : ℂ) ^ s / 2)
+    (completedCosZeta a s) := by
+  let c (n : ℤ) : ℂ := cexp (2 * π * I * a * n) / 2
+  have hF t (ht : 0 < t) : HasSum (fun n : ℤ ↦ if n = 0 then 0 else c n * rexp (-π * n ^ 2 * t))
+      ((cosKernel a t - 1) / 2) := by
+    refine ((hasSum_int_cosKernel₀ a ht).div_const 2).congr_fun fun n ↦ ?_
+    split_ifs with h <;>
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:515]: `with h` seems unnecessary.

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 5]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+      RCLike.norm_ofNat, norm_norm, Complex.norm_exp_ofReal_mul_I, add_zero, norm_one,
+      norm_of_nonneg (by positivity : 0 ≤ |(i : ℝ)| ^ s.re), div_right_comm, le_rfl]
+  · simp only [c, Int.cast_eq_zero, ← Int.cast_abs, ofReal_intCast, div_right_comm, mul_div_assoc]
+
+/-- Formula for `completedCosZeta` as a Dirichlet series in the convergence range
+(second version, with sum over `ℕ`). -/
+lemma hasSum_nat_completedCosZeta (a : ℝ) {s : ℂ} (hs : 1 < re s) :
+    HasSum (fun n : ℕ ↦ if n = 0 then 0 else Gammaℝ s * Real.cos (2 * π * a * n) / (n : ℂ) ^ s)
+    (completedCosZeta a s) := by
+  have hs' : s ≠ 0 := (not_lt.mpr zero_le_one <| zero_re ▸ · ▸ hs)
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:536]: This is [Complex.ne_zero_of_one_lt_re](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Complex/Abs.html#Complex.ne_zero_of_one_lt_re).

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 6]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+  simpa only [one_div, sub_self, div_zero, Gammaℝ_one, div_one, sub_zero] using
+    (differentiableAt_hurwitzZetaEven_sub_one_div a).continuousAt.tendsto
+
+lemma differentiable_hurwitzZetaEven_sub_hurwitzZetaEven (a b : UnitAddCircle) :
+    Differentiable ℂ (fun s ↦ hurwitzZetaEven a s - hurwitzZetaEven b s) := by
+  intro z
+  rcases ne_or_eq z 1 with hz | rfl
+  · exact (differentiableAt_hurwitzZetaEven a hz).sub (differentiableAt_hurwitzZetaEven b hz)
+  · -- NB. This can be written more tidy with `convert`, but the `convert` version is 3x slower, as
+    -- it spends 877 TC synthesis steps vainly trying to infer `Subsingleton (ℂ → ℂ)`.
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:681]: Hopefully this problem will go away once [lean4#3996](https://github.com/leanprover/lean4/issues/3996) has been dealt with.

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 7]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+        1 / (s - 1) / Gammaℝ s) - (hurwitzZetaEven b s - 1 / (s - 1) / Gammaℝ s)) := by abel
+    exact funext this ▸ (differentiableAt_hurwitzZetaEven_sub_one_div a).sub
+      (differentiableAt_hurwitzZetaEven_sub_one_div b)
+
+/--
+Formula for `hurwitzZetaEven` as a Dirichlet series in the convergence range, with sum over `ℤ`.
+-/
+lemma hasSum_int_hurwitzZetaEven (a : ℝ) {s : ℂ} (hs : 1 < re s) :
+    HasSum (fun n : ℤ ↦ 1 / (↑|n + a| : ℂ) ^ s / 2) (hurwitzZetaEven a s) := by
+  rw [hurwitzZetaEven, Function.update_noteq (not_lt.mpr zero_le_one <| zero_re ▸ · ▸ hs)]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:692]: This looks like [Complex.ne_zero_of_one_lt_re](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Complex/Abs.html#Complex.ne_zero_of_one_lt_re) again (there may be other instances I have overlooked...).

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 8]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+    HasSum (fun n : ℤ ↦ cexp (2 * π * I * a * n) / ↑|n| ^ s / 2) (cosZeta a s) := by
+  rw [cosZeta, Function.update_noteq (not_lt.mpr zero_le_one <| zero_re ▸ · ▸ hs)]
+  refine ((hasSum_int_completedCosZeta a hs).div_const (Gammaℝ s)).congr_fun fun n ↦ ?_
+  rw [mul_div_assoc _ (cexp _), div_right_comm _ (2 : ℂ),
+    mul_div_cancel_left₀ _ (Gammaℝ_ne_zero_of_re_pos (zero_lt_one.trans hs))]
+
+/-- Formula for `cosZeta` as a Dirichlet series in the convergence range, with sum over `ℕ`. -/
+lemma hasSum_nat_cosZeta (a : ℝ) {s : ℂ} (hs : 1 < re s) :
+    HasSum (fun n : ℕ ↦ Real.cos (2 * π * a * n) / (n : ℂ) ^ s) (cosZeta a s) := by
+  have hs' : s ≠ 0 := (fun h ↦ (not_lt.mpr zero_le_one) ((zero_re ▸ h ▸ hs)))
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:767]: See above.

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 9]
```diff
@@ -395,3 +407,406 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+  simp_rw [push_cast, Complex.cos, neg_mul]
+  exact this.congr_fun fun n ↦ by rw [show 2 * π * a * n * I = 2 * π * I * a * n by ring]
+
+/-- Reformulation of `hasSum_nat_cosZeta` using `LSeriesHasSum`. -/
+lemma LSeriesHasSum_cos (a : ℝ) {s : ℂ} (hs : 1 < re s) :
+    LSeriesHasSum (Real.cos <| 2 * π * a * ·) s (cosZeta a s) := by
+  refine (hasSum_nat_cosZeta a hs).congr_fun (fun n ↦ ?_)
+  rcases eq_or_ne n 0 with rfl | hn
+  · rw [LSeries.term_zero, Nat.cast_zero, Nat.cast_zero,
+      zero_cpow (fun h ↦ (not_lt.mpr zero_le_one) ((zero_re ▸ h ▸ hs))), div_zero]
```
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:781]: Another one?

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean (hunk 7 above)
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:693]: I fixed the two you found and one other – I think that's the lot.

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean (hunk 6 above)
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:682]: I added a TODO note.

Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean]: With the latest mathlib master, the `convert` is now fast again, so I have reinstated it.

**Comment by MichaelStollBayreuth** [comment]
LGTM.
maintainer merge

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean [hunk 10]
```diff
@@ -395,3 +407,398 @@ lemma completedCosZeta₀_one_sub (a : UnitAddCircle) (s : ℂ) :
...
+  · simp_rw [← mul_one_div ‖_‖]
+    apply Summable.mul_left
+    rwa [summable_one_div_int_add_rpow]
+  · rw [mul_one_div, div_right_comm]
+
+/-!
+## The un-completed even Hurwitz zeta
+-/
+
+/-- Technical lemma which will give us differentiability of Riemann zeta at `s = 0`. -/
```
Comment (sgouezel) [inline Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:569]: Are you planning to apply this to streamline something about Riemann zeta, or should one just change the docstring by replacing Riemann with Hurwitz?

**Review by sgouezel — COMMENTED** [review]
bors d+

File: Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean (hunk 10 above)
Comment (loefflerd) [reply Mathlib/NumberTheory/LSeries/HurwitzZetaEven.lean:569]: Good point, that was unclear, I'll change it.

**Comment by loefflerd** [comment]
bors r+

(I know it is normal to wait for CI to pass before sending a delegated PR to bors, but here, the only change since delegation is a one-word edit to a doc comment, so there is no possibility of CI not passing.)

**Comment by loefflerd** [comment]
I'm mystified why the bors run failed. I've just merged master and it builds both locally and in CI, so let's try again

bors r+

**Pull request closed without merging**